### ISSUE TITLE
SaveParticipantTest.testFormatChange* failed with additional whitespace

### ISSUE THREAD

**Issue by jukzi** [eclipse-jdt/eclipse.jdt.ui#1445]
SaveParticipantTest

4 fails in https://download.eclipse.org/eclipse/downloads/drops4/I20240608-0320/testresults/html/org.eclipse.jdt.ui.tests_ep433I-unit-mac64-java17_macosx.cocoa.x86_64_17.html
for example:

```
expected:<... { String s [= (String) ]o; } }> but was:<... { String s [ = (String)]o; } }>

org.junit.ComparisonFailure: expected:<... {
String s [= (String) ]o;
}
}> but was:<... {
String s [ = (String)]o;
}
}>
at org.junit.Assert.assertEquals(Assert.java:117)
at org.junit.Assert.assertEquals(Assert.java:146)
at org.eclipse.jdt.ui.tests.quickfix.SaveParticipantTest.testFormatChanges01(SaveParticipantTest.java:152)
at java.base/java.lang.reflect.Method.invoke(Method.java:568)
```

![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/51790620/6cfec0c5-b76f-429c-84e0-b0e053685d4a)

**Comment by jukzi** [comment]
Locally **randomly** reproducible sometimes on  various test methods. examples:
![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/51790620/8ff87ca2-66e7-4bc8-8823-af3471faf02a)
![image](https://github.com/eclipse-jdt/eclipse.jdt.ui/assets/51790620/554126e3-d6db-489e-90cd-e285759ac58e)

**Comment by jukzi** [comment]
again on mac:
https://download.eclipse.org/eclipse/downloads/drops4/I20240619-1800/testresults/html/org.eclipse.jdt.ui.tests_ep433I-unit-mac64-java17_macosx.cocoa.x86_64_17.html

**Comment by jukzi** [comment]
duplicate: https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/79 (also only mac)

**Comment by jukzi** [comment]
single test fail on linux:
https://download.eclipse.org/eclipse/downloads/drops4/I20240619-1800/testresults/html/org.eclipse.jdt.ui.tests_ep433I-unit-cen64-gtk3-java21_linux.gtk.x86_64_21.html
```
expected:<...* A block comment on[] two lines */ ...> but was:<...* A block comment on[ * ] two lines */ ...>

org.junit.ComparisonFailure: expected:<...* A block comment on[] two lines
*/

...> but was:<...* A block comment on[
* ] two lines
*/

...>
at org.junit.Assert.assertEquals(Assert.java:117)
at org.junit.Assert.assertEquals(Assert.java:146)
at org.eclipse.jdt.ui.tests.quickfix.SaveParticipantTest.testBug232768_2(SaveParticipantTest.java:790)
at java.base/java.lang.reflect.Method.invoke(Method.java:580)
```

**Comment by carstenartur** [comment]
I think that is an old one. A number of reasons are possible. One could be that the formatting and compiler settings data structure gets duplicated in some situations. So working on a copy in some threads is not thread save. When changes are applied at runtime the question is which set of options is changed. That is a bug in jdt core. That is from the top of my head - could be in the meantime things have been improved. And then there was an issue with the undo not sure if related. I think in one of the junit tests you might still find a sleep to stabilize results.

**Comment by jukzi** [comment]
https://download.eclipse.org/eclipse/downloads/drops4/I20250115-1800/testresults/html/org.eclipse.jdt.ui.tests_ep435I-unit-win32-x86_64-java21_win32.win32.x86_64_21.html

```
org.junit.ComparisonFailure: expected:<...;
public class E1 {
[ /**
* Method foo with a really long description that will wrap lines on save
* operation
*
* @param a
* - integer input
*
* @return integer
*/
public int foo(int a) {
return 0;
]}
}> but was:<...;
public class E1 {
[ /**
* Method foo with a really long description that will wrap lines on save
* operation
*
* @param a - integer input
*
* @return integer
*/
public int foo(int a) {
return 0;
]}
}>
at org.junit.Assert.assertEquals(Assert.java:117)
at org.junit.Assert.assertEquals(Assert.java:146)
at org.eclipse.jdt.ui.tests.quickfix.SaveParticipantTest.testFormatChangeBug561164(SaveParticipantTest.java:1105)
at java.base/java.lang.reflect.Method.invoke(Method.java:580)
```

**Comment by carstenartur** [comment]
This is an old one. I still wonder if it might be caused by jdt.core to have more than one instance of an preferences settings set under some circumstances.

https://bugs.eclipse.org/bugs/show_bug.cgi?id=570172

**Comment by carstenartur** [comment]
Maybe duplicate of

https://bugs.eclipse.org/bugs/show_bug.cgi?id=360711

**Comment by carstenartur** [comment]
This points to another related bug

https://bugs.eclipse.org/bugs/show_bug.cgi?id=563771

**Comment by jukzi** [comment]
> https://bugs.eclipse.org/bugs/show_bug.cgi?id=563771

i don't know how a performance thing would relate. However a problem with JavaModelManager.getOptions() is, that it returns a modifiable Hashtable but  the returned copy should not be used to modify the (cached) map. A clean solution would return a Hashtable generated like Collections.unmodifiableMap().

**Comment by jukzi** [comment]
debugged: GapTextStore is created in another thread (main) then it is read (JavaReconciler). As this Datastructur is not threadsafe that is wrong

**Comment by jukzi** [comment]
> JavaModelManager.getOptions()
> A clean solution would return a Hashtable generated like Collections.unmodifiableMap().
Or better to be compatible a thing like a CopyOnWriteHashtable. @carstenartur do you want to work on that?

**Comment by jukzi** [comment]
In a case where a test failed locally i found this Exception(s) in my console, which is not shown when the tests succeed:
```
!ENTRY org.eclipse.jdt.ui 4 10006 2025-01-23 17:32:42.836
!MESSAGE The save participant 'org.eclipse.jdt.ui.postsavelistener.cleanup' caused an exception: org.eclipse.core.runtime.CoreException: End position lies outside document range
!STACK 1
org.eclipse.core.runtime.CoreException: End position lies outside document range
	at org.eclipse.ltk.internal.core.refactoring.Changes.asCoreException(Changes.java:66)
	at org.eclipse.ltk.core.refactoring.TextChange.perform(TextChange.java:247)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpChange.perform(CleanUpRefactoring.java:167)
	at org.eclipse.ltk.core.refactoring.PerformChangeOperation.lambda$0(PerformChangeOperation.java:247)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2457)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2482)
	at org.eclipse.ltk.core.refactoring.PerformChangeOperation.executeChange(PerformChangeOperation.java:284)
	at org.eclipse.ltk.core.refactoring.PerformChangeOperation.run(PerformChangeOperation.java:213)
	at org.eclipse.jdt.internal.corext.fix.CleanUpPostSaveListener.saved(CleanUpPostSaveListener.java:407)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitDocumentProvider$4.run(CompilationUnitDocumentProvider.java:1610)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitDocumentProvider.notifyPostSaveListeners(CompilationUnitDocumentProvider.java:1604)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitDocumentProvider.commitWorkingCopy(CompilationUnitDocumentProvider.java:1388)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitDocumentProvider$3.execute(CompilationUnitDocumentProvider.java:1463)
	at org.eclipse.ui.editors.text.TextFileDocumentProvider$DocumentProviderOperation.run(TextFileDocumentProvider.java:133)
	at org.eclipse.ui.actions.WorkspaceModifyDelegatingOperation.execute(WorkspaceModifyDelegatingOperation.java:71)
	at org.eclipse.ui.actions.WorkspaceModifyOperation.lambda$0(WorkspaceModifyOperation.java:112)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2457)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2482)
	at org.eclipse.ui.actions.WorkspaceModifyOperation.run(WorkspaceModifyOperation.java:131)
	at org.eclipse.ui.internal.editors.text.WorkspaceOperationRunner.run(WorkspaceOperationRunner.java:76)
	at org.eclipse.ui.internal.editors.text.WorkspaceOperationRunner.run(WorkspaceOperationRunner.java:66)
	at org.eclipse.ui.editors.text.TextFileDocumentProvider.executeOperation(TextFileDocumentProvider.java:461)
	at org.eclipse.ui.editors.text.TextFileDocumentProvider.saveDocument(TextFileDocumentProvider.java:783)
	at org.eclipse.ui.texteditor.AbstractTextEditor.performSave(AbstractTextEditor.java:5020)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitEditor.performSave(CompilationUnitEditor.java:1262)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitEditor.doSave(CompilationUnitEditor.java:1316)
	at org.eclipse.jdt.ui.tests.quickfix.SaveParticipantTest.editCUInEditor(SaveParticipantTest.java:75)
	at org.eclipse.jdt.ui.tests.quickfix.SaveParticipantTest.testFormatChangeBug488229_3(SaveParticipantTest.java:1032)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:198)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:169)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:93)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:58)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:141)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:57)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:103)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:94)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:52)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:70)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:100)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:520)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:748)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:443)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:83)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable.lambda$1(E4Testable.java:127)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4077)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3693)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1151)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1042)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:665)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:573)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:178)
	at org.eclipse.pde.internal.junit.runtime.NonUIThreadTestApplication.start(NonUIThreadTestApplication.java:58)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:208)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:143)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:109)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:439)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:271)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:668)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:605)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1481)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1454)
Caused by: org.eclipse.text.edits.MalformedTreeException: End position lies outside document range
	at org.eclipse.text.edits.TextEditProcessor.checkIntegrityDo(TextEditProcessor.java:180)
	at org.eclipse.text.edits.TextEdit.dispatchCheckIntegrity(TextEdit.java:746)
	at org.eclipse.text.edits.TextEditProcessor.performEdits(TextEditProcessor.java:153)
	at org.eclipse.ltk.core.refactoring.TextChange.performEdits(TextChange.java:278)
	at org.eclipse.ltk.core.refactoring.TextFileChange.access$0(TextFileChange.java:1)
	at org.eclipse.ltk.core.refactoring.TextFileChange.lambda$0(TextFileChange.java:280)
	at org.eclipse.ui.internal.editors.text.UISynchronizationContext.run(UISynchronizationContext.java:35)
	at org.eclipse.core.internal.filebuffers.TextFileBufferManager.execute(TextFileBufferManager.java:593)
	at org.eclipse.ltk.core.refactoring.TextFileChange.performEdits(TextFileChange.java:291)
	at org.eclipse.ltk.core.refactoring.TextChange.perform(TextChange.java:239)
	... 96 more
!SUBENTRY 1 org.eclipse.ltk.core.refactoring 4 10001 2025-01-23 17:32:42.838
!MESSAGE End position lies outside document range
!STACK 0
org.eclipse.text.edits.MalformedTreeException: End position lies outside document range
	at org.eclipse.text.edits.TextEditProcessor.checkIntegrityDo(TextEditProcessor.java:180)
	at org.eclipse.text.edits.TextEdit.dispatchCheckIntegrity(TextEdit.java:746)
	at org.eclipse.text.edits.TextEditProcessor.performEdits(TextEditProcessor.java:153)
	at org.eclipse.ltk.core.refactoring.TextChange.performEdits(TextChange.java:278)
	at org.eclipse.ltk.core.refactoring.TextFileChange.access$0(TextFileChange.java:1)
	at org.eclipse.ltk.core.refactoring.TextFileChange.lambda$0(TextFileChange.java:280)
	at org.eclipse.ui.internal.editors.text.UISynchronizationContext.run(UISynchronizationContext.java:35)
	at org.eclipse.core.internal.filebuffers.TextFileBufferManager.execute(TextFileBufferManager.java:593)
	at org.eclipse.ltk.core.refactoring.TextFileChange.performEdits(TextFileChange.java:291)
	at org.eclipse.ltk.core.refactoring.TextChange.perform(TextChange.java:239)
	at org.eclipse.jdt.internal.corext.fix.CleanUpRefactoring$CleanUpChange.perform(CleanUpRefactoring.java:167)
	at org.eclipse.ltk.core.refactoring.PerformChangeOperation.lambda$0(PerformChangeOperation.java:247)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2457)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2482)
	at org.eclipse.ltk.core.refactoring.PerformChangeOperation.executeChange(PerformChangeOperation.java:284)
	at org.eclipse.ltk.core.refactoring.PerformChangeOperation.run(PerformChangeOperation.java:213)
	at org.eclipse.jdt.internal.corext.fix.CleanUpPostSaveListener.saved(CleanUpPostSaveListener.java:407)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitDocumentProvider$4.run(CompilationUnitDocumentProvider.java:1610)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitDocumentProvider.notifyPostSaveListeners(CompilationUnitDocumentProvider.java:1604)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitDocumentProvider.commitWorkingCopy(CompilationUnitDocumentProvider.java:1388)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitDocumentProvider$3.execute(CompilationUnitDocumentProvider.java:1463)
	at org.eclipse.ui.editors.text.TextFileDocumentProvider$DocumentProviderOperation.run(TextFileDocumentProvider.java:133)
	at org.eclipse.ui.actions.WorkspaceModifyDelegatingOperation.execute(WorkspaceModifyDelegatingOperation.java:71)
	at org.eclipse.ui.actions.WorkspaceModifyOperation.lambda$0(WorkspaceModifyOperation.java:112)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2457)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2482)
	at org.eclipse.ui.actions.WorkspaceModifyOperation.run(WorkspaceModifyOperation.java:131)
	at org.eclipse.ui.internal.editors.text.WorkspaceOperationRunner.run(WorkspaceOperationRunner.java:76)
	at org.eclipse.ui.internal.editors.text.WorkspaceOperationRunner.run(WorkspaceOperationRunner.java:66)
	at org.eclipse.ui.editors.text.TextFileDocumentProvider.executeOperation(TextFileDocumentProvider.java:461)
	at org.eclipse.ui.editors.text.TextFileDocumentProvider.saveDocument(TextFileDocumentProvider.java:783)
	at org.eclipse.ui.texteditor.AbstractTextEditor.performSave(AbstractTextEditor.java:5020)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitEditor.performSave(CompilationUnitEditor.java:1262)
	at org.eclipse.jdt.internal.ui.javaeditor.CompilationUnitEditor.doSave(CompilationUnitEditor.java:1316)
	at org.eclipse.jdt.ui.tests.quickfix.SaveParticipantTest.editCUInEditor(SaveParticipantTest.java:75)
	at org.eclipse.jdt.ui.tests.quickfix.SaveParticipantTest.testFormatChangeBug488229_3(SaveParticipantTest.java:1032)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:198)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:169)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:93)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:58)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:141)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:57)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:103)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:94)
	at org.junit.platform.launcher.core.DelegatingLauncher.execute(DelegatingLauncher.java:52)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:70)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:100)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:520)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:748)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:443)
	at org.eclipse.pde.internal.junit.runtime.RemotePluginTestRunner.main(RemotePluginTestRunner.java:83)
	at org.eclipse.pde.internal.junit.runtime.PlatformUITestHarness.lambda$0(PlatformUITestHarness.java:45)
	at org.eclipse.e4.ui.internal.workbench.swt.E4Testable.lambda$1(E4Testable.java:127)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:132)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4077)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3693)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1151)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1042)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:665)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:573)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:178)
	at org.eclipse.pde.internal.junit.runtime.NonUIThreadTestApplication.start(NonUIThreadTestApplication.java:58)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:208)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:143)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:109)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:439)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:271)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:668)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:605)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1481)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1454)

!ENTRY org.eclipse.jface 4 4 2025-01-23 17:32:42.870
!MESSAGE Save Problems : The file has been saved but one of the save actions failed.

```

**Comment by jukzi** [comment]
> In a case where a test failed locally i found this Exception(s) in my console

in another case there was no exception in the console

**Comment by carstenartur** [comment]
> > https://bugs.eclipse.org/bugs/show_bug.cgi?id=563771
> 
> i don't know how a performance thing would relate. However a problem with JavaModelManager.getOptions() is, that it returns a modifiable Hashtable but  the returned copy should not be used to modify the (cached) map. A clean solution would return a Hashtable generated like Collections.unmodifiableMap().

The multithreading issue in the original version was meant to be fixed by creating a copy of the options map in some situations. This made the exception go away but introduced the performance issue while multithreading still is wrong.

**Comment by trancexpress** [comment]
`testFormatChangesBug208568` is still failing, e.g.: https://download.eclipse.org/eclipse/downloads/drops4/I20260826-2300/testresults/html/org.eclipse.jdt.ui.tests_ep441I-unit-linux-x86_64-java25_linux.gtk.x86_64_25.html

```
contents was not formated. Actual: package test1; public class E1 { public int i= 10; }

java.lang.AssertionError: contents was not formated. Actual: package test1;
public class E1 {
public int i= 10;

}

at org.junit.Assert.fail(Assert.java:89)
at org.junit.Assert.failEquals(Assert.java:187)
at org.junit.Assert.assertNotEquals(Assert.java:163)
at org.eclipse.jdt.ui.tests.quickfix.SaveParticipantTest.assertChangedFromTo(SaveParticipantTest.java:999)
at org.eclipse.jdt.ui.tests.quickfix.SaveParticipantTest.testFormatChangesBug208568(SaveParticipantTest.java:506)
at java.base/java.lang.reflect.Method.invoke(Method.java:565) 
```

I see this running in parallel:

```
| Worker-8: Reading saved version of file for quick diff  org.eclipse.text | /debug | org.eclipse.jface.text.CopyOnWriteTextStore | set | 154 | CopyOnWriteTextStore.set() text=package test1;
public class E1 {
    public  int i= 10;

}
```

The job itself locks the document with a rule, see `org.eclipse.ui.internal.editors.quickdiff.LastSaveReferenceProvider.readDocument(IProgressMonitor, boolean)`:

```
	private void readDocument(IProgressMonitor monitor, boolean force) {
            [..]
				ISchedulingRule rule= getSchedulingRule(storage);
                                [..]
				try {
					lockDocument(monitor, jobMgr, rule);
```

We could try try locking in the test as well, to avoid collisions...

```
		IStorageEditorInput input = (IStorageEditorInput) editor.getEditorInput();
		ISchedulingRule rule= input.getStorage().getAdapter(ISchedulingRule.class);
		Job.getJobManager().beginRule(rule, null);
		try {
			cu.getBuffer().setContents(newContent);
			editor.doSave(null);
		} finally {
			Job.getJobManager().endRule(rule);
		}
```

But that doesn't guarantee the job won't run after we locked the document, potentially overwriting the test contents.

So I guess we try waiting on the job, hopefully that helps.

**Comment by carstenartur** [comment]
> > JavaModelManager.getOptions()
> > A clean solution would return a Hashtable generated like Collections.unmodifiableMap().
> 
> Or better to be compatible a thing like a CopyOnWriteHashtable. @carstenartur do you want to work on that?
> 
> 

Hi @jukzi, thanks for the suggestion! I’ve now followed up on the options-cache side and opened [eclipse-jdt/eclipse.jdt.core#5364](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/5364).

The investigation led to a slightly different approach than a `CopyOnWriteHashtable`: an in-flight reader can overwrite an updated or invalidated cache with an older snapshot. The proposed fix keeps the existing defensive copies and uses generation-checked cache publication to prevent this.

The PR includes deterministic regression tests. A separate integration test with the real Java editor also demonstrates that this race can cause incorrect cast spacing during a save. The relevant race assertions fail without the fix and pass with it. However, I haven’t reproduced the complete original failure reported here, including the extra space before `=`, so I’m not claiming that this resolves #1445 in full.

Would you be willing to take a look when you have a chance? I’d particularly appreciate your feedback on the synchronization approach and the regression tests, as well as any thoughts on the remaining gap to the original failure. Thanks for your help!
